PSReadLine/Render.cs

```diff
@@ -378,13 +378,21 @@ void UpdateColorsIfNecessary(string newColor)
 
                     // promptBufferCells is the number of visible characters in the prompt
                     var promptBufferCells = LengthInBufferCells(promptText);
-                    _console.CursorLeft -= promptBufferCells;
-                    var color = renderData.errorPrompt ? _options._errorColor : defaultColor;
-                    if (renderData.errorPrompt && promptBufferCells != promptText.Length)
-                        promptText = promptText.Substring(promptText.Length - promptBufferCells);
-                    UpdateColorsIfNecessary(color);
-                    _console.Write(promptText);
-                    _console.Write("\x1b[0m");
+
+                    // The 'CursorLeft' could be less than error prompt cell length when
+                    //   1. console buffer was resized and the regular prompt takes a whole line of the new width buffer; OR
+                    //   2. the prompt function was changed and the new prompt is shorter than the error prompt.
+                    // when this happens, we skip changing the color of the prompt.
+                    if (_console.CursorLeft >= promptBufferCells)
```

Comment:
It seems like this could be confusing - e.g. the prompt text would stay red even though it shouldn't be anymore.

I'm also surprised that line 368 works correctly after resize. In fact, that seems like a bigger problem and part of the necessary fix here to correctly change the prompt color back.

In the second case, that seems like possibly something to warn the user about? The `PromptText` option should normally be a very short string.

Reply:
> I'm also surprised that line 368 works correctly after resize

The x and y are adjusted in `RecomputeInitialCoords()`, and that's why line 368 works correctly.

> part of the necessary fix here to correctly change the prompt color back

Hmm, I don't think this will be a problem when the above 2 conditions happen.
If the prompt color was changed to red previously, that means the editing is still in the same logical line, and thus `_initialX` and `_initialY` will not be changed, namely `_console.CursorLeft` will not change.

> In the second case, that seems like possibly something to warn the user about?

Maybe. There are actually a couple of tiny problems when the prompt is changed:
1. the error prompt is not changed, so when there is any parsing error, we will end up changing the prompt text. For example, prompt is changed to `PS>` while error prompt is still `> `, then the prompt displayed will be `P>>` when there is a parsing error.
2. we don't know the prompt is changed unless we monitor the prompt function (obj reference equality? definition text equality?), and we will have to get the error prompt again based on the new prompt string.
My take is that changing prompt function within a session should be very rare and it's not worth the effort to track the changes and make things look naturally good when it's changed. What's your thoughts?

Reply:
Isn't this code hit only when the color needs to change, either to red or back to normal (line 375?)

Assuming that, not doing it will be misleading, but it's possible it was already misleading, in other words, a color change was missed already?

If the cursor is placed correctly after resize, you know the cursor is immediately after the prompt. Moving back is slightly more complicated when you have to span lines, but it's not hard.

As for knowing about the prompt change - the idea is to only color the end of the prompt - we infer the last non-whitespace character plus the trailing whitespace (so we can move the cursor backwards correctly).

Right now, if we can't infer what to color, it's the users responsibility to configure PSReadLine correctly. It would be nice to warn when it's not configured properly, but I don't think that's easy, and maybe not worth the complexity.

Reply:
> Isn't this code hit only when the color needs to change, either to red or back to normal (line 375?)
Assuming that, not doing it will be misleading, but it's possible it was already misleading, in other words, a color change was missed already?

I thought changing color from red back to normal won't be affected even after resizing the width buffer, since once the color has been successfully changed to red, it can be changed back to normal only if the same logical lines are being edited, and in that case, `_initialX` and `_initialY` won't be changed, even if a change in the buffer width happens. It turns out I was wrong at one edge case -- take `ps C:\Users\UserA\AppData\Local\Microsoft\ApplicationInsights> 'abcde` as an example, if the buffer width shrank to the point that the very first  single quote `'` gets shuffled to the next line, then `_initialX` and `_initialY` get changed (something seems wrong in this case), and a closing single quote will trigger the `ArgumentOutOfRangeException`. In this particular case, the red `>` from the above line won't be changed back to normal with changes in this PR.

I will redo the fix to handle coloring in the "buffer resizing" situation.
But for the "prompt change" situation, I think it makes sense to just skip coloring, unless we want to go with the warning.

> It would be nice to warn when it's not configured properly, but I don't think that's easy, and maybe not worth the complexity.

Same thoughts here.

Reply:
@lzybkr Is there a way to check if the current cursor is on a wrapped line (it's actually on the same line as the previous line and just got wrapped into a new line because the line is longer than the width buffer)?
I believe that knowledge is needed in order to decide if the `CursorLeft` should be changed to point to the end of the previous line.
After executing a command, `_initialX` could become less than `promptBufferCells` for a few reasons:
1. console buffer got resized, in which case we would know before executing the command
2. executing the command causes the prompt to be longer just enough to cause the line to be wrapped.
3. the `prompt` function is changed to produce a very short prompt string, say 1 character.

Ideally, we should adjust `CursorLeft` to the end of previous line only for case 1 and 2, where the cursor is actually on the same line (wrapped). For case 3, we should just skip changing the color.

I could evaluate the prompt function in case `_console.CursorLeft >= promptBufferCells`, so as to decide if the cursor is on a wrapped line, but I really don't want to do that if there is another way to find out this.

Reply:
I forget the details, but the logic in `ReallyRender` needs the same sort of thing - that's the difference between a "physical line" (a line as rendered in the console) and a logical line (which, when long, might span more than 1 physical line.) It's mostly simple logic, the first line is special cased because of the prompt, otherwise it's simple arithmetic between the character (or in your case cursor position) and the buffer width.

Note that in versions <2.0, we never needed to reevaluate the prompt because we could scrape the console buffer. That's no longer possible in a portable way, but I've avoided reevaluating the prompt because people do sometimes display things like the time, and it could be misleading if the time changes while editing the command line. There are of course times when we must run the prompt again, but that happens either because the user asked for it (redraw over the current prompt), or after showing possible completions where the prompt needs to be drawn *below* those completions.

Reply:
The `logical line` and `physical line` calculation in `ReallyRender` doesn't help here.
The existing `logical line` and `physical line` calculation is simple because we have all the text that are going to be rendered. But here, we need to decide if the current cursor is within the same line as the previous line, which has already been rendered on the console (by PowerShell itself writing out the prompt string). We don't have the text at hand to calculate if the current cursor is on a wrapped line or a new line.

The only way I can think of is to reevaluate the prompt function **without rendering it on the console**, then we have the text and can do the calculation to decide if the current cursor is on the wrapped line or the new line. Do you think this approach is acceptable?

Reply:
You can't evaluate the prompt to capture it's output. Some prompts write some/much of their content directly, e.g. via `Write-Host` so they can get colors.

Are you just wanting `ConvertOffsetToPoint`?

Reply:
About the prompt evaluation, you are right, no guarantee to capture its exact output. So that approach won't work.

`ConvertOffsetToPoint` is not what I want. It's useful for calculating cursor position but won't help to determine if the cursor is in the wrapped line of the previous line.

Here is an example where the cursor should be adjusted to the end of the previous physical line to change the color of `>`, because even though the cursor is actually in the wrapped line (same logical line).

![image](https://user-images.githubusercontent.com/127450/61488882-2f792800-a95e-11e9-8d9c-e1437930a624.png)

Here is an example where the cursor shouldn't be adjusted because it's literally on a new line, not the wrapped line of the previous physical line.

![image](https://user-images.githubusercontent.com/127450/61489089-bf1ed680-a95e-11e9-881d-9f661f01103d.png)

The problem is I have difficulty to tell which cases we are in when `_console.CursorLeft >= promptBufferCells`. Monitoring the change of `prompt` function doesn't help too, because there is no way to tell it's changed to produce a very short string -- it's possible the change of the `prompt` function produces a longer prompt string that causes the line to be wrapped.

Reply:
> You can't evaluate the prompt to capture it's output. Some prompts write some/much of their content directly, e.g. via `Write-Host` so they can get colors.

Since PSRL only supports 5.1+ you can handle that:

```csharp
using (var ps = PowerShell.Create(RunspaceMode.CurrentRunspace))
{
    var command = new Command("profile", isScript: false, useLocalScope: true);
    command.MergeMyResults(PipelineResultTypes.Information, PipelineResultTypes.Output);
    ps.Commands.AddCommand(command);
    var includesInfoRecords = ps.Invoke();
}
```

Accounting for things like escape sequences and `\b` would add a lot of complexity though.

Reply:
IMO, fixing the color changing of `>` in this edge case doesn't worth the effort.
If we accept this PR as it is, then in the scenarios shown as above screenshots, what would happen is just the prompt's color won't change when there is any parsing errors. This is fine, as we don't change the color anyways when the prompt function [is not `pure`](https://github.com/PowerShell/PSReadLine/blob/master/PSReadLine/ReadLine.cs#L746-L765).

The worst case would be when the user started by typing on the same line:
![image](https://user-images.githubusercontent.com/127450/61552586-154a5300-aa0d-11e9-8aba-40bb12eb99a7.png)

And then, the user carefully resizes the window to make the typed text appear on the next line:
![image](https://user-images.githubusercontent.com/127450/61552687-580c2b00-aa0d-11e9-8d21-fc1a1e91413c.png)

Then in this case, the existing red `>` won't be changed back to normal when the parsing error is gone. But it affects that particular prompt line only.
I think this is acceptable, as the user needs to carefully crafted it to make this happen.

Reply:
It sounds like it might be hit somewhat frequently though, otherwise how did you end up finding this? Or did you find it from the short prompt and then realized this case as well?

Do note that `PromptText` is user configurable for non-trivial prompts.

It bothers me because somewhat sticky (you'd have to introduce another error then fix that before it is correct.)  I think people will <kbd>Ctrl+c</kbd> if they notice assuming the prompt is broken.

At any rate, isn't the correct fix to simply wrap to the previous line as though you were pressing left arrow?

Reply:
> It sounds like it might be hit somewhat frequently though, otherwise how did you end up finding this? Or did you find it from the short prompt and then realized this case as well?

I'm not sure which case you are referring to here.
I went through all the `ArgumentOutOfRangeException` issues with the `Parameter name: left` being the problem, and they can be categorized in to 3 buckets. Please see the PR description for the details.
Basically, users run into this with 3 conditions:
1. Resize the windows, so the cursor appears in the next line (1 issue)
2. Prompt string gets longer, and cause the cursor to appear in the next line. (3+ issues)
3. Prompt function is changed and the new prompt string is too short. (3+ issues)

None of them result in the "worst case" as shown in https://github.com/PowerShell/PSReadLine/pull/967#discussion_r305447896. I found the "worst case" when I started to really play with some edge cases.

> At any rate, isn't the correct fix to simply wrap to the previous line as though you were pressing left arrow?

If we know we should do this, then it's simple to do. The problem is we don't know if we should do this.
See the screenshots in https://github.com/PowerShell/PSReadLine/pull/967#discussion_r305100779, when the prompt function is changed to produce a short prompt string, we shouldn't wrap to the previous line.
This is why I asked if there is a way to tell whether the cursor is on a wrapped line. So far, I don't find an easy way to determine it.
@SeeminglyScience had a suggestion at https://github.com/PowerShell/PSReadLine/pull/967#discussion_r305322159, but my personal opinion is the value doesn't worth the complexity.

Reply:
Thanks, this makes more sense now.

So for 1 & 2, assuming we wrapped is correct.
For 3, assuming we wrapped is wrong.

If we always assume we wrapped and fix appropriately, we:
* avoid an exception
* get the color correct for cases 1 & 2
* get fewer bug reports
* in case 3, rendering might be off

We can also argue case 3 is user error - and that they need to update PromptText. I think we might see fewer bug reports too, especially if we publish a sample recipe for changing the prompt (like I have a `Push-Prompt` and `Pop-Prompt` pair of functions.

Is this reasonable?

Reply:
This sounds reasonable. Great, it's a much easier problem if we don't need to accurately differentiate 1,2 from 3.